### PR TITLE
Add versioning by URL

### PR DESCRIPTION
**Pull request by frankharkins**
This PR implements versioning, accessible by URL (works towards #1844).

### Notebooks directory structure

This PR adds another directory level for notebooks covered by version control. For example, the `ch-algorithms` folder has moved to two separate folders (`v1` & `v2`), but `summer-schools` is not covered so remains unchanged.

```diff
 platypus
 └── notebooks
     ├── summer-schools
     ├── toc.yaml
-    └── ch-algorithms
-        └── shor.ipynb
+    ├── v1
+    │   └── ch-algorithms
+    │       └── shor.ipynb
+    └── v2
+        └── ch-algorithms
+            └── shor.ipynb
```

Then `notebooks/toc.yaml` must point to these new notebooks (475d1009b74ef34ee2eae80e32f1892072096770). We also add a prefix (e.g. `v1_`) to the url for the server to identify versioned folders.

```diff
# notebooks/toc.yaml
 - title: Quantum Protocols and Quantum Algorithms
-  url: /ch-algorithms
+  url: /v1_ch-algorithms
   sections:
     - title: Shor's Algorithm
-      url: ch-algorithms/shor
+      url: v1/notebooks/ch-algorithms/shor
```

### Server changes

The above updates to the `notebooks` directory adds URLs
```
/course/v1_ch-algorithms/shor
/course/v2_ch-algorithms/shor
```
and removes the URL
```
/course/ch-algorithms/shor
```

For better user experience, this PR adapts the server to recognise the following URLs and show the appropriate content:

| Accessed URL | Content shown |
|:---|:---|
| `/v1/course/ch-algorithms/shor` | `/course/v1_ch-algorithms/shor` |
| `/v2/course/ch-algorithms/shor` | `/course/v2_ch-algorithms/shor` |
| `/course/ch-algorithms/shor` | `/course/v2_ch-algorithms/shor` |

### Other

I've updated the documentation, and included a small script I used to add the new versions to the TOC.

## How to read this PR

This PR has a huge diff as 51b7f27edb0d12fc05a377909d4f7fd30a5237ae removes all textbook-related notebooks and replaces with submodules, so probably best to check things out commit-wise.

- 62794869bdfa69677a86c5007f06bff35b5da924 and 005db299987f761893ad90f7dc55a9595b2cb35a are logic changes to the converter and server
- 51b7f27edb0d12fc05a377909d4f7fd30a5237ae and 475d1009b74ef34ee2eae80e32f1892072096770 are content changes (with huge diffs)
- 1a3c52aeae8972ec71ac9063fd8ea1d9c003b5ef updates documentation

Testing locally, you may need to run `git submodule update` to pull the content from Qiskit/textbook.

EDIT: [Preview link](https://platypus-pr-1845.ecy0akwlcpw.us-south.codeengine.appdomain.cloud/course/ch-states/introduction) now available.

## Screenshots

Video below shows two different versions of the same page (noticeable difference is the lack of annotations on the old version).

https://user-images.githubusercontent.com/36071638/214296427-3379ea05-e65c-40c3-b3f6-a6189ca2fa99.mov

## Future work

Out of scope of this PR:

- Add banner to tell users they're viewing an old version of the textbook
- Translations

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,6 @@
+[submodule "notebooks/v1"]
+	path = notebooks/v1
+	url = https://github.com/Qiskit/textbook.git
+[submodule "notebooks/v2"]
+	path = notebooks/v2
+	url = https://github.com/Qiskit/textbook.git
```
Comment (vabarbosa) [inline .gitmodules:6]: just curious should the branch be included in these somewhere?

**Review by vabarbosa — CHANGES_REQUESTED** [review]
@frankharkins this is great! addition to the couple minor comments (below), i think you will need to 

- update the [Dockerfile](https://github.com/Qiskit/platypus/blob/main/Dockerfile#L20-L21) to initialize the submodules

- update the [release action](https://github.com/Qiskit/platypus/blob/main/.github/workflows/release.yml#L34-L36) to initialize the submodules

- make the textbook repo public so the preview and build will be able to clone them. otherwise, will need update the preview & build scripts to use a specific tokens with access to the repo. (i believe this is the reason for the [preview failing](https://github.com/Qiskit/platypus/actions/runs/3997380194/jobs/6858632839#step:11:137))

let me know if you have questions or need assistance with any of these.

lastly looking at the[ textbook repo](https://github.com/Qiskit/textbook) should we have the specific branches also tagged and release created? that way they would be called out in the front page and also any wanting a specific version could just download the release instead of dealing with cloning the repo, etc. (just a thought).

File: .gitmodules (hunk 1 above)
Comment (frankharkins) [reply .gitmodules:7]: Oops, fixed in [9a6448c](https://github.com/Qiskit/platypus/pull/1845/commits/9a6448ce7894d939f4f35b245c8dbef5714a979c) - thanks

File: server/utilities.ts [hunk 2]
```diff
@@ -25,6 +25,8 @@ import { IS_PRODUCTION } from './configuration'
 
 const TEXTBOOK_HOME = 'https://qiskit.org/learn'
 
+const LATEST_TEXTBOOK_VERSION = 'v2'
```
[the remark this answers is not in the document]
Comment (frankharkins) [reply server/utilities.ts:28]: Yes, thanks! Fixed in [fbd65e9](https://github.com/Qiskit/platypus/pull/1845/commits/fbd65e9eec5739617104d4739a95e885c2739766)

**Comment by vabarbosa** [comment]
@frankharkins, the preview has now been created. however, it looks like the pages are not available. for example,

https://platypus-pr-1845.ecy0akwlcpw.us-south.codeengine.appdomain.cloud/course/introduction/why-quantum-computing

when i have the opportunity i will try to review why this afternoon

**Comment by frankharkins** [comment]
Whoops looks like the action is skipped the submodule update, hopefully should work now.

**Comment by vabarbosa** [comment]
> Whoops looks like the action is skipped the submodule update, hopefully should work now.

the latest preview is looking much better now. thank you. i will go ahead and approve gh actions PR (and then revert [this](https://github.com/Qiskit/platypus/pull/1845/commits/7d8be6b6d3eb1253571576015d0b99f5ebf80c84))

**Review by vabarbosa — APPROVED** [review]
@frankharkins  the preview was successful 
https://platypus-pr-1845.ecy0akwlcpw.us-south.codeengine.appdomain.cloud/course/introduction/why-quantum-computing
and all looks good to me. **thank you!!**

the one other change i would suggest is to clean up the package.json and remove the `test:nb:*` related scripts (if no longer needed)

thanks again!

**Pull request merged**